File: src/claude_bridge/notify.py
"""Task completion notification delivery."""

import json
import os
from datetime import datetime
from urllib.request import urlopen, Request

from .db import BridgeDB


def get_bot_token() -> str | None:
    """Get Telegram bot token from environment or config."""
    token = os.environ.get("TELEGRAM_BOT_TOKEN")
    if token:
        return token

    # Try config file
    config_path = os.path.expanduser("~/.claude-bridge/config.json")
    if os.path.isfile(config_path):
        try:
            with open(config_path) as f:
                config = json.load(f)
            return config.get("telegram_bot_token")
        except (json.JSONDecodeError, IOError):
            pass

    # Try .mcp.json in common bridge-bot locations
    for path in [
        os.path.expanduser("~/projects/bridge-bot/.mcp.json"),
        os.path.expanduser("~/projects/claude-bridge-bot/.mcp.json"),
    ]:
        if os.path.isfile(path):
            try:
                with open(path) as f:
                    mcp = json.load(f)
                token = (mcp.get("mcpServers", {})
                         .get("telegram", {})
                         .get("env", {})
                         .get("TELEGRAM_BOT_TOKEN"))
                if token:
                    return token
            except (json.JSONDecodeError, IOError):
                pass

    return None


def format_completion_message(task, agent_name: str) -> str:
    """Format a task completion message for notification."""
    task_id = task["id"]
    status = task["status"]
    prompt = task["prompt"][:80].split("\n")[0]
    task_type = task["task_type"] or "standard"

    duration = ""
    if task["duration_ms"]:
        mins = task["duration_ms"] // 60000
        secs = (task["duration_ms"] % 60000) // 1000
        duration = f"{mins}m {secs}s"

    cost = f"${task['cost_usd']:.3f}" if task["cost_usd"] else ""

    if status == "done":
        icon = "🏁" if task_type == "team" else "✓"
        lines = [f"{icon} Task #{task_id} ({agent_name}) — done"]
        if duration:
            lines[0] += f" in {duration}"
        summary = task["result_summary"] or ""
        if summary:
            lines.append(summary[:200])
        if cost:
            lines.append(f"Cost: {cost}")
    else:
        lines = [f"✗ Task #{task_id} ({agent_name}) — {status}"]
        if duration:
            lines[0] += f" after {duration}"
        error = task["error_message"] or ""
        if error:
            lines.append(f"Error: {error[:200]}")
        if cost:
            lines.append(f"Cost: {cost}")

    return "\n".join(lines)


def send_telegram(bot_token: str, chat_id: str, message: str) -> bool:
    """Send a message via Telegram Bot API. Returns True on success."""
    url = f"https://api.telegram.org/bot{bot_token}/sendMessage"
    payload = json.dumps({"chat_id": chat_id, "text": message}).encode()
    req = Request(url, data=payload, headers={"Content-Type": "application/json"})
    try:
        with urlopen(req, timeout=10) as resp:
            body = resp.read()
            result = json.loads(body)
            return result.get("ok", False)
    except Exception:
        return False


def deliver_notification(db: BridgeDB, notification_id: int) -> bool:
    """Attempt to deliver a notification. Returns True on success."""
    notif = db.get_notification(notification_id)
    if not notif or notif["status"] != "pending":
        return False

    channel = notif["channel"]
    if channel == "telegram":
        token = get_bot_token()
        if not token:
            return False
        success = send_telegram(token, notif["chat_id"], notif["message"])
    else:
        # Other channels not yet implemented
        return False

    if success:
        db.mark_notification_sent(notification_id)
    return success

File: src/execution/notify.ts
/**
 * Notifier — sends task completion notifications to Telegram.
 *
 * Output is rendered as Telegram HTML so Markdown that Claude emits in
 * `result_summary` (e.g. `**bold**`, fenced code, `## headings`) renders
 * properly instead of leaking literal markup.
 */

import { existsSync, readFileSync } from "fs";
import { join } from "path";
import type { Task, Notification } from "../types.js";
import type { INotifier } from "./interfaces.js";
import { TelegramFormatter } from "../channel/telegram/format.js";
import { mdToTelegramHtml } from "../channel/telegram/markdown-to-html.js";

// TelegramFormatter is used internally for HTML-aware chunking via splitMessage().

// Telegram caps sendMessage at 4096 UTF-16 code units. Stay under to leave room
// for the `(i/n)` chunk header and emoji surrogate pairs.
const TELEGRAM_CHUNK_LIMIT = 3900;

export class Notifier implements INotifier {
  constructor(private homeDir: string) {}

  /**
   * Compose a task-completion message in Markdown. `notify()` converts to HTML
   * before sending, so this lives in the same Markdown vocabulary as every
   * other Notifier caller (bridge_reply, on-complete summary, loop status).
   */
  formatMessage(task: Task, agentName: string): string {
    const ok = task.status === "done";
    const icon = ok ? "✅" : "❌";
    const statusLabel = ok ? "done" : task.status;

    const headerParts: string[] = [`${icon} **${agentName}**`];
    headerParts.push(`_${statusLabel}_`);
    if (task.prompt) {
      const shortPrompt = task.prompt.length > 120
        ? task.prompt.slice(0, 120) + "…"
        : task.prompt;
      headerParts.push("`" + shortPrompt.replace(/`/g, "'") + "`");
    }

    const parts: string[] = [headerParts.join(" · ")];

    if (ok && task.result_summary) {
      // Quote each line so the converter wraps the body in <blockquote>.
      const quoted = task.result_summary
        .split("\n")
        .map((l) => `> ${l}`)
        .join("\n");
      parts.push("", quoted);
    }

    if (!ok && task.error_message) {
      parts.push("", `> ${task.error_message}`);
    }

    const meta: string[] = [];
    if (task.cost_usd !== null) meta.push(`$${task.cost_usd.toFixed(2)}`);
    if (task.duration_ms !== null) {
      const secs = Math.round(task.duration_ms / 1000);
      meta.push(secs >= 60 ? `${Math.floor(secs / 60)}m${secs % 60}s` : `${secs}s`);
    }
    if (task.num_turns !== null) meta.push(`${task.num_turns} turns`);
    if (meta.length > 0) parts.push("", `_${meta.join(" · ")}_`);

    return parts.join("\n");
  }

  getBotToken(): string | null {
    const configPath = join(this.homeDir, "config.json");
    if (existsSync(configPath)) {
      try {
        const config = JSON.parse(readFileSync(configPath, "utf-8")) as Record<string, unknown>;
        if (config["telegram_token"]) {
          return config["telegram_token"] as string;
        }
      } catch { /* ignore */ }
    }
    return process.env["TELEGRAM_BOT_TOKEN"] ?? null;
  }

  /**
   * Split a message into chunks no larger than `limit` UTF-16 code units.
   * HTML-aware: keeps `<pre>` / `<blockquote>` blocks intact and re-wraps them
   * across boundaries when an oversized block must be split.
   */
  static splitMessage(text: string, limit: number = TELEGRAM_CHUNK_LIMIT): string[] {
    if (text.length <= limit) return [text];
    return new TelegramFormatter().chunkMessage(text, limit);
  }

  async notify(notification: Pick<Notification, "chat_id" | "message">): Promise<boolean> {
    const token = this.getBotToken();
    if (!token) return false;

    // All callers (bridge_reply, on-complete summary, loop status) send Markdown.
    // Convert once here so `**bold**` / fenced code / lists render properly under
    // parse_mode: HTML instead of leaking literal markup.
    const html = mdToTelegramHtml(notification.message);
    const chunks = Notifier.splitMessage(html);
    const total = chunks.length;

    for (let i = 0; i < total; i++) {
      const header = total > 1 ? `<i>(${i + 1}/${total})</i>\n` : "";
      const text = `${header}${chunks[i]}`;
      const ok = await this.sendOne(token, notification.chat_id, text, i + 1, total);
      if (!ok) return false;
    }
    return true;
  }

  private async sendOne(
    token: string,
    chatId: string | number,
    text: string,
    part: number,
    total: number,
  ): Promise<boolean> {
    // Try HTML first; fall back to plain text if Telegram rejects entity parsing.
    const html = await this.postMessage(token, chatId, text, "HTML");
    if (html.ok) return true;

    if (html.parseError) {
      process.stderr.write(
        `[notify] HTML parse rejected by Telegram, falling back to plain text. ` +
        `chat=${chatId} part=${part}/${total} desc="${html.desc.slice(0, 200)}"\n`,
      );
      const plain = await this.postMessage(token, chatId, stripHtml(text), undefined);
      if (plain.ok) return true;
      this.logFailure(chatId, part, total, plain, text.length);
      return false;
    }

    this.logFailure(chatId, part, total, html, text.length);
    return false;
  }

  private async postMessage(
    token: string,
    chatId: string | number,
    text: string,
    parseMode: "HTML" | undefined,
  ): Promise<PostResult> {
    try {
      const body: Record<string, unknown> = {
        chat_id: chatId,
        text,
        disable_web_page_preview: true,
      };
      if (parseMode) body["parse_mode"] = parseMode;

      const resp = await fetch(`https://api.telegram.org/bot${token}/sendMessage`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(body),
      });
      const raw = await resp.text().catch(() => "");
      let json: { ok?: boolean; description?: string; error_code?: number } = {};
      try { json = JSON.parse(raw); } catch { /* non-JSON */ }
      const desc = json.description ?? raw;
      const httpOk = resp.ok && json.ok === true;
      return {
        ok: httpOk,
        status: resp.status,
        code: json.error_code,
        desc,
        parseError:
          !httpOk && /can't parse|can not parse|unsupported start tag|unexpected/i.test(desc),
      };
    } catch (err) {
      return {
        ok: false,
        status: 0,
        code: undefined,
        desc: err instanceof Error ? err.message : String(err),
        parseError: false,
      };
    }
  }

  private logFailure(
    chatId: string | number,
    part: number,
    total: number,
    res: PostResult,
    len: number,
  ): void {
    process.stderr.write(
      `[notify] Telegram http=${res.status} code=${res.code ?? "?"} ` +
      `chat=${chatId} part=${part}/${total} len=${len}: ${res.desc.slice(0, 300)}\n`,
    );
  }

  async retryFailed(): Promise<void> {
    // Will be implemented when integrated with DB in later waves
  }
}

interface PostResult {
  ok: boolean;
  status: number;
  code: number | undefined;
  desc: string;
  parseError: boolean;
}

/** Best-effort HTML stripper for the plain-text fallback path. */
function stripHtml(s: string): string {
  return s
    .replace(/<\/?(?:b|i|u|s|code|pre|blockquote|tg-spoiler)(?:\s[^>]*)?>/gi, "")
    .replace(/<a\s+[^>]*href="([^"]*)"[^>]*>([\s\S]*?)<\/a>/gi, "$2 ($1)")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&quot;/g, '"')
    .replace(/&amp;/g, "&");
}
